### PR TITLE
feat: make ZNSRegistry a UUPS proxy

### PR DESCRIPTION
**Pull request by JamesEarle**

### CODE REVIEW

File: contracts/access/AccessControlled.sol [hunk 1]
```diff
@@ -2,20 +2,34 @@
...
 
     IZNSAccessController internal accessController;
 
-
     modifier onlyAdmin() {
         accessController.checkAdmin(msg.sender);
         _;
     }
 
+    modifier onlyGovernor() {
```
Comment (Whytecrowe) [inline contracts/access/AccessControlled.sol]: probably don't need these. It will overload the size of every contract. we have these on access controller already. should be enough

File: contracts/registry/ZNSRegistry.sol [hunk 2]
```diff
@@ -45,17 +42,12 @@ contract ZNSRegistry is AccessControlled, ERC1967UpgradeUpgradeable, IZNSRegistr
         _;
     }
 
-    modifier onlyRegistrar() {
```
Comment (Whytecrowe) [inline contracts/registry/ZNSRegistry.sol:48]: leave this

File: contracts/registry/ZNSRegistry.sol [hunk 3]
```diff
@@ -201,11 +193,11 @@ contract ZNSRegistry is AccessControlled, ERC1967UpgradeUpgradeable, IZNSRegistr
 
     function setAccessController(
         address accessController
-    ) external override(AccessControlled, IZNSRegistry) onlyAdmin {
```
Comment (Whytecrowe) [inline contracts/registry/ZNSRegistry.sol:204]: why did you remove this?

File: contracts/registry/IZNSRegistry.sol [hunk 4]
```diff
@@ -152,12 +152,4 @@ interface IZNSRegistry {
      * @param domainHash The hash of the domain name
      */
     function deleteRecord(bytes32 domainHash) external;
-
-    /**
-     * @notice Set the access controller contract
-     * @param accessController The new access controller
-     */
-    function setAccessController(address accessController) external;
```
Comment (Whytecrowe) [inline contracts/registry/IZNSRegistry.sol:160]: why did you remove these? they will be inaccessible, no?

Comment (Whytecrowe) [reply contracts/registry/IZNSRegistry.sol:160]: dangerous thing you are doing. please, let me know why

File: test/ZNSRegistry.test.ts [hunk 5]
```diff
@@ -275,34 +260,34 @@ describe("ZNSRegistry Tests", () => {
...
-      const tx = registry.updateDomainRecord(domainHash, mockResolver.address, ethers.constants.AddressZero);
+      await zns.registry.connect(mockRegistrar).createDomainRecord(domainHash, deployer.address, mockResolver.address);
+      const tx = zns.registry.updateDomainRecord(domainHash, mockResolver.address, ethers.constants.AddressZero);
 
       await expect(tx).to.be.fulfilled;
     });
 
-    it("Cannot update a domain owner if owner is zero address", async () => {
-      const tx = registry
+    it("Cannot set a domain owner if owner is zero address", async () => {
```
Comment (Whytecrowe) [inline test/ZNSRegistry.test.ts]: why are you changing the wording back? I specifically changed to "update" cause that's what we're doing. the function on Registry is called `upadateDomainOwner()`. why making it more confusing?
Update means changing existing, set means setting a value in any case. "set" is very confusing here

File: test/ZNSRegistry.test.ts [hunk 6]
```diff
@@ -312,21 +297,21 @@ describe("ZNSRegistry Tests", () => {
       await expect(tx).to.be.revertedWith(OWNER_NOT_ZERO_REG_ERR);
     });
 
-    it("Can update a domain resolver if resolver is zero address", async () => {
-      await registry
+    it("Can set a domain resolver if resolver is zero address", async () => {
```
Comment (Whytecrowe) [inline test/ZNSRegistry.test.ts]: it's **update**DomainResolver()

File: test/ZNSRegistry.test.ts [hunk 7]
```diff
@@ -312,21 +297,21 @@ describe("ZNSRegistry Tests", () => {
...
 
-      const zeroResolver = await registry.getDomainResolver(wilderDomainHash);
+      const zeroResolver = await zns.registry.getDomainResolver(wilderDomainHash);
 
       expect(zeroResolver).to.be.eq(ethers.constants.AddressZero);
     });
 
-    it("Fails to set a record when caller is not owner", async () => {
-      const tx = registry
+    it("Fails to set a record when caller is not owner or operator", async () => {
```
Comment (Whytecrowe) [inline test/ZNSRegistry.test.ts]: "update" should be here as well

File: test/ZNSRegistry.test.ts [hunk 8]
```diff
@@ -459,20 +449,151 @@ describe("ZNSRegistry Tests", () => {
...
+        accessControllerAfter,
+      ] = await registryPromises(zns.registry, domainHash, [deployer, operator]);
+
+      expect(ownerStatusBefore).to.eq(ownerStatusAfter);
+      expect(operatorStatusBefore).to.eq(operatorStatusAfter);
+      expect(existanceBefore).to.eq(existanceAfter);
+      expect(accessControllerBefore).to.eq(accessControllerAfter);
+    });
+
+    it("Fails when an unauthorized account tries to call to upgrade", async () => {
```
Comment (Whytecrowe) [inline test/ZNSRegistry.test.ts]: this tests's title contradicts the test itself

File: test/helpers/checkUpgrades.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  ];
+
+  return Promise.all(calls);
+};
+
+export const upgradedRegistryActions = async (
+  upgradedRegistry : ZNSRegistryMock,
+  newNumber : BigNumber,
+  randomUser : SignerWithAddress
+) => {
```
Comment (Whytecrowe) [inline test/helpers/checkUpgrades.ts]: maybe we can change the helpers to a different style:
we can make these more abstract to work with any contract, and this will check values from arg1 to the values from arg2, both of which can be contract specific callbacks that we pass as an array for each arg.
we specify the calls as args in the test itself, this way it will be easily readable of what exactly we are checking in the test, while the helper will make these calls and do comparisons internally. we don't need to see comparisons in the tests as much as what we are checking exactly, which we can tell by the callbacks we pass to the helper in the test body.
what do you think?

File: contracts/registry/IZNSRegistry.sol (hunk 4 above)
Comment (JamesEarle) [reply contracts/registry/IZNSRegistry.sol:160]: @Whytecrowe what makes this dangerous? We are inheriting it twice, once from the `AccessControlled` contract and then again in the interface for this contract. I was removing the second reference as it seems unnecessary to do this twice, Why do we need this? It's already `virtual` in the `AccessControlled` contract we inherit from so there isn't a possibility of not overriding it as it will error if we don't

File: test/ZNSRegistry.test.ts (hunk 5 above)
Comment (JamesEarle) [reply test/ZNSRegistry.test.ts]: moved back to update

File: test/ZNSRegistry.test.ts (hunk 8 above)
Comment (JamesEarle) [reply test/ZNSRegistry.test.ts]: I had mistakenly put the negative test case in the positive test and vice versa, I've already reversed this in the latest however

File: contracts/registry/ZNSRegistry.sol (hunk 3 above)
Comment (JamesEarle) [reply contracts/registry/ZNSRegistry.sol:204]: See below comment, what is the need for inheriting this twice if it's already enforced by `AccessControlled` ?

File: contracts/access/AccessControlled.sol [hunk 10]
```diff
@@ -2,15 +2,14 @@
 pragma solidity ^0.8.18;
 
 import { IZNSAccessController } from "./IZNSAccessController.sol";
-
+import { ZNSRoles } from "./ZNSRoles.sol";
```
Comment (Whytecrowe) [inline contracts/access/AccessControlled.sol]: why is this imported? we don't need this anymore

File: contracts/distribution/IZNSEthRegistrar.sol [hunk 11]
```diff
@@ -28,22 +28,30 @@ interface IZNSEthRegistrar {
...
-    function setAddressResolver(address znsAddressResolver_) external;
+    function setAddressResolver(address addressResolver_) external;
 
     function setAccessController(address accessController_) external;
 
     function getAccessController() external view returns (address);
+
+    function initialize(
+        address accessController_,
+        address znsRegistry_,
```
Comment (Whytecrowe) [inline contracts/distribution/IZNSEthRegistrar.sol:52]: should probably remove "zns" from here as well then

File: contracts/distribution/IZNSEthRegistrar.sol [hunk 12]
```diff
@@ -28,22 +28,30 @@ interface IZNSEthRegistrar {
...
+        address znsTreasury_,
+        address znsDomainToken_,
+        address znsAddressResolver_
```
Comment (Whytecrowe) [inline contracts/distribution/IZNSEthRegistrar.sol]: should probably remove "zns" from here as well then

File: contracts/distribution/IZNSTreasury.sol [hunk 13]
```diff
@@ -52,4 +52,11 @@ interface IZNSTreasury {
     function setAccessController(address accessController) external;
 
     function getAccessController() external view returns (address);
+
+    function initialize(
+        address accessController_,
+        address znsPriceOracle_,
```
Comment (Whytecrowe) [inline contracts/distribution/IZNSTreasury.sol]: same here

File: contracts/distribution/ZNSEthRegistrar.sol [hunk 14]
```diff
@@ -94,14 +93,12 @@ contract ZNSEthRegistrar is AccessControlled, IZNSEthRegistrar {
...
-    function revokeDomain(bytes32 domainHash)
-    external
-    override
     // TODO: figure out how to guard this so people can stake tokens
     //  without the risk of staking contract or wallet to call reclaim+revoke
     //  from underneath them
+    function revokeDomain(bytes32 domainHash) external override 
     onlyNameOwner(domainHash)
-    onlyTokenOwner(domainHash)
+    onlyTokenOwner(domainHash)
```
Comment (Whytecrowe) [inline contracts/distribution/ZNSEthRegistrar.sol]: why space?

File: contracts/distribution/ZNSEthRegistrar.sol [hunk 15]
```diff
@@ -111,60 +108,56 @@ contract ZNSEthRegistrar is AccessControlled, IZNSEthRegistrar {
         emit DomainRevoked(domainHash, msg.sender);
     }
 
-    function reclaimDomain(bytes32 domainHash)
-    external
-    override
+    function reclaimDomain(bytes32 domainHash) external override
     onlyTokenOwner(domainHash)
```
Comment (Whytecrowe) [inline contracts/distribution/ZNSEthRegistrar.sol]: if we newline, let's newlineeverything. here only modifier is newlined. let's keep to the recommended syntax for Solidity - revert back to previous or put everything in one line.

File: contracts/distribution/ZNSEthRegistrar.sol [hunk 16]
```diff
@@ -111,60 +108,56 @@ contract ZNSEthRegistrar is AccessControlled, IZNSEthRegistrar {
...
-        emit AddressResolverSet(znsAddressResolver_);
+        emit AddressResolverSet(addressResolver_);
     }
 
-    function setAccessController(address accessController_)
-    external
-    override(AccessControlled, IZNSEthRegistrar)
-    onlyAdmin
+    function setAccessController(address accessController_) external override(AccessControlled, IZNSEthRegistrar) 
+        onlyAdmin
```
Comment (Whytecrowe) [inline contracts/distribution/ZNSEthRegistrar.sol]: again here. why did you bring the visibility keywords up, but not the modifier? what's wrong with the conventional approach that's already been there? also, why is modifier indented? this is "dirty syntax", let's revert back to how it was.
and also, let's not do this again, please. we're wasting time reformatting the same thing over and over and commenting on it. there's a reason why I formatted it this way, no need to change it

File: contracts/distribution/ZNSPriceOracle.sol [hunk 17]
```diff
@@ -77,6 +77,8 @@ contract ZNSPriceOracle is AccessControlled, Initializable, IZNSPriceOracle {
 
         fee = getRegistrationFee(domainPrice);
         totalPrice = domainPrice + fee;
+
+        return (totalPrice, domainPrice, fee);
```
Comment (Whytecrowe) [inline contracts/distribution/ZNSPriceOracle.sol]: let's do one or the other method. and not half both. either remove this line or specify naming locally. imo, it was already fine

File: contracts/distribution/ZNSTreasury.sol [hunk 18]
```diff
@@ -63,11 +61,11 @@ contract ZNSTreasury is AccessControlled, IZNSTreasury {
...
-      // Transfer stake amount and fee
-      stakingToken.transferFrom(depositor, address(this), stakeAmount);
-      // TODO make sure we show the approval process to the user here to avoid failed transfer
-      // TODO can we make it so it needs a single approval only?!
-      stakingToken.transferFrom(depositor, zeroVault, registrationFee);
+        // Transfer stake amount and fee
+        stakingToken.transferFrom(depositor, address(this), stakeAmount);
+        // TODO make sure we show the approval process to the user here to avoid failed transfer
+        // TODO can we make it so it needs a single approval only?!
+        stakingToken.transferFrom(depositor, zeroVault, registrationFee);
```
Comment (Whytecrowe) [inline contracts/distribution/ZNSTreasury.sol]: remove whitespace at the end

File: contracts/registry/IZNSRegistry.sol (hunk 4 above)
Comment (Whytecrowe) [reply contracts/registry/IZNSRegistry.sol:160]: either remove everywhere or leave this one as well.

File: contracts/registry/ZNSRegistry.sol [hunk 19]
```diff
@@ -131,7 +128,7 @@ contract ZNSRegistry is AccessControlled, ERC1967UpgradeUpgradeable, IZNSRegistr
         bytes32 domainHash,
         address owner,
         address resolver
-    ) external override onlyRegistrar {
+    ) external override onlyRegistrar() {
```
Comment (Whytecrowe) [inline contracts/registry/ZNSRegistry.sol]: no parentheses needed

File: contracts/registry/ZNSRegistry.sol (hunk 3 above)
Comment (Whytecrowe) [reply contracts/registry/ZNSRegistry.sol:204]: see the previous comment

File: contracts/registry/ZNSRegistry.sol [hunk 20]
```diff
@@ -201,11 +198,11 @@ contract ZNSRegistry is AccessControlled, ERC1967UpgradeUpgradeable, IZNSRegistr
 
     function setAccessController(
         address accessController
-    ) external override(AccessControlled, IZNSRegistry) onlyAdmin {
+    ) external override onlyAdmin {
         _setAccessController(accessController);
     }
 
-    function getAccessController() external view override(AccessControlled, IZNSRegistry) returns (address) {
+    function getAccessController() external view override returns (address) {
```
Comment (Whytecrowe) [inline contracts/registry/ZNSRegistry.sol]: same here

File: contracts/token/IZNSDomainToken.sol [hunk 21]
```diff
@@ -1,12 +1,32 @@
...
+    /**
+     * @notice Mints a token with a specified tokenId, using _safeMint, and sends it to the given address
+     * @param to The address that will recieve the newly minted domain token
+     * @param tokenId The TokenId that the caller wishes to mint/register
+     */
     function register(address to, uint256 tokenId) external;
 
+    /**
+     * @notice Burns the token with the specified tokenId
+     * @dev TODO: Add onlyRole(REGISTRAR_ROLE)
```
Comment (Whytecrowe) [inline contracts/token/IZNSDomainToken.sol]: outdated comment

File: contracts/token/ZNSDomainToken.sol [hunk 22]
```diff
@@ -1,27 +1,27 @@
...
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
 import { IZNSDomainToken } from "./IZNSDomainToken.sol";
 import { AccessControlled } from "../access/AccessControlled.sol";
 
 
 /**
  * @title A contract for tokenizing domains under ZNS
  */
-contract ZNSDomainToken is AccessControlled, ERC721, IZNSDomainToken {
+contract ZNSDomainToken is AccessControlled, UUPSUpgradeable, ERC721Upgradeable, IZNSDomainToken {
```
Comment (Whytecrowe) [inline contracts/token/ZNSDomainToken.sol:12]: think on the order of inheritance here. do we want all the uups slots be before all the token slots? what about the gaps?

File: contracts/token/ZNSDomainToken.sol [hunk 23]
```diff
@@ -1,27 +1,27 @@
...
-
-    constructor(
+    function initialize(
+        address accessController,
         string memory tokenName,
-        string memory tokenSymbol,
-        address accessController
-    ) ERC721(tokenName, tokenSymbol) {
+        string memory tokenSymbol
+    ) public override initializer {
```
Comment (Whytecrowe) [inline contracts/token/ZNSDomainToken.sol]: why it's public here, but external in other contracts? let's bring them to the same visibility everywhere

File: contracts/token/ZNSDomainToken.sol [hunk 24]
```diff
@@ -38,11 +38,11 @@ contract ZNSDomainToken is AccessControlled, ERC721, IZNSDomainToken {
      * @notice Burns the token with the specified tokenId
      * @param tokenId The tokenId that the caller wishes to burn/revoke
      */
-    function revoke(uint256 tokenId) external override onlyRegistrar {
+    function revoke(uint256 tokenId) external override onlyRegistrar  {
```
Comment (Whytecrowe) [inline contracts/token/ZNSDomainToken.sol]: extra space

File: contracts/token/ZNSDomainToken.sol [hunk 25]
```diff
@@ -38,11 +38,11 @@ contract ZNSDomainToken is AccessControlled, ERC721, IZNSDomainToken {
...
         _burn(tokenId);
     }
 
-    function setAccessController(address accessController)
+    function setAccessController(address accessController)
```
Comment (Whytecrowe) [inline contracts/token/ZNSDomainToken.sol]: extra space

File: contracts/upgradeMocks/Mock.sol [hunk 26]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+ /* solhint-disable */
+contract Mock {
```
Comment (Whytecrowe) [inline contracts/upgradeMocks/Mock.sol]: let's maybe make the naming more clear of what this "Mock" is. specify that it's for the upgrade tests

File: contracts/upgradeMocks/distribution/ZNSEthRegistrarMock.sol [hunk 27]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import { ZNSEthRegistrar } from "../../distribution/ZNSEthRegistrar.sol";
+import { Mock } from "../Mock.sol";
+
+ /* solhint-disable */
+contract ZNSEthRegistrarMock is ZNSEthRegistrar, Mock {}
```
Comment (Whytecrowe) [inline contracts/upgradeMocks/distribution/ZNSEthRegistrarMock.sol]: same here. let's add "upgrade" to the name or something that clarifies what this is, so that it's clear why these mocks exist. for all the below contracts as well

File: package.json [hunk 28]
```diff
@@ -11,21 +11,22 @@
...
+    "lint": "yarn lint-sol & yarn lint-ts",
     "clean": "hardhat clean",
     "build": "yarn run clean && yarn run compile",
     "typechain": "hardhat typechain",
     "test": "hardhat test",
     "semantic-release": "semantic-release --tag-format='v${version}-dev'",
-    "coverage": "hardhat coverage",
-    "check-coverage": "istanbul check-coverage --statements 90 --branches 87 --functions 89 --lines 90"
+    "check-coverage": "istanbul check-coverage --statements 90 --branches 87 --functions 89 --lines 90",
+    "coverage": "hardhat coverage && yarn check-coverage"
```
Comment (Whytecrowe) [inline package.json]: why did you combine it? did you look into circleci yaml config?
we need these to be different scripts because they are different jobs in the CI. since you've added && here, it will not go through fully if one of the scripts fails, breaking our flow in CircleCI.
we also need to be able to get local coverage files without checking coverage, so that we can actually see what tests we are missing.
let's bring them back to be separate. and before making changes always ensure you are not breaking anything else by doing that.

File: test/ZNSAddressResolver.test.ts [hunk 29]
```diff
@@ -1,195 +1,232 @@
...
   });
 
   it("Should not allow non-owner address to setAddress", async () => {
     await expect(
-      znsAddressResolver.connect(addr1).setAddress(wilderDomainNameHash, addr1.address)
-    ).to.be.revertedWith(
-      "ZNSAddressResolver: Not authorized for this domain"
-    );
+      zns.addressResolver.connect(user).setAddress(wilderDomainHash, user.address)
+    ).to.be.revertedWith("ZNSAddressResolver: Not authorized for this domain");
```
Comment (Whytecrowe) [inline test/ZNSAddressResolver.test.ts:107]: why do you keep reformatting my code for small things like this? how many times will we format back and forth? :)
maybe we should create a guide or something and stick to it, if you are still not satisfied with how I format. Thought we've talked about this already :)
this is all just extra time for small things like this...

File: test/ZNSAddressResolver.test.ts [hunk 30]
```diff
@@ -1,195 +1,232 @@
...
 
     await expect(
-      znsAddressResolver.connect(mockRegistrar)
-        .setAddress(wilderDomainNameHash, addr1.address)
-    )
-      .to.emit(znsAddressResolver, "AddressSet")
-      .withArgs(wilderDomainNameHash, addr1.address);
-  });
+      zns.addressResolver.connect(mockRegistrar).setAddress(wilderDomainHash, hre.ethers.constants.AddressZero)
+    ).to.emit(zns.addressResolver, "AddressSet").withArgs(wilderDomainHash, hre.ethers.constants.AddressZero);
```
Comment (Whytecrowe) [inline test/ZNSAddressResolver.test.ts]: again here. this looks dirty and hard to read. why did you reformat? how does this help?

File: test/ZNSDomainToken.test.ts [hunk 31]
```diff
@@ -136,13 +148,74 @@ describe("ZNSDomainToken:", () => {
...
+    it("Fails to upgrade if the caller is not authorized", async () => {
+      // UUPS specifies that a call to upgrade must be made through an address that is upgradecall
+      // So use a deployed proxy contract
+      const factory = new ZNSDomainTokenMock__factory(deployer);
+
+      // DomainToken to upgrade to
+      const newDomainToken = await factory.deploy();
+      await newDomainToken.deployed();
+
+      // Confirm the deployer is a governor, as set in `deployZNS` helper
```
Comment (Whytecrowe) [inline test/ZNSDomainToken.test.ts]: wrong comment

File: test/ZNSEthRegistrar.test.ts [hunk 32]
```diff
@@ -16,27 +17,26 @@ import { priceConfigDefault } from "./helpers/constants";
...
 require("@nomicfoundation/hardhat-chai-matchers");
 
-
 describe("ZNSEthRegistrar", () => {
   let deployer : SignerWithAddress;
   let user : SignerWithAddress;
   let governor : SignerWithAddress;
   let admin : SignerWithAddress;
-  let randomAcc : SignerWithAddress;
+  let randomUser : SignerWithAddress;
```
Comment (Whytecrowe) [inline test/ZNSEthRegistrar.test.ts:31]: :)

File: test/ZNSEthRegistrar.test.ts [hunk 33]
```diff
@@ -527,101 +530,158 @@ describe("ZNSEthRegistrar", () => {
...
+
+    it("Fails to upgrade when an unauthorized users calls", async () => {
+      const registrarFactory = new ZNSEthRegistrar__factory(deployer);
+      const registrar = await registrarFactory.deploy();
+      await registrar.deployed();
+
+      const tx = zns.registrar.connect(randomUser).upgradeTo(registrar.address);
+
+      await expect(tx).to.be.revertedWith(
+        `AccessControl: account ${randomUser.address.toLowerCase()} is missing role ${GOVERNOR_ROLE}`
```
Comment (Whytecrowe) [inline test/ZNSEthRegistrar.test.ts]: use helper here

File: test/ZNSRegistry.test.ts (hunk 6 above)
Comment (Whytecrowe) [reply test/ZNSRegistry.test.ts:297]: this was resolved, however nothing has changed

File: test/helpers/deployZNS.ts [hunk 34]
```diff
@@ -61,27 +73,40 @@ export const deployPriceOracle = async ({
...
+    }
+  ) as ZNSPriceOracle;
 
   return priceOracle;
 };
 
 export const deployDomainToken = async (
   deployer : SignerWithAddress,
-  accessControllerAddress : string
+  accessController : string
```
Comment (Whytecrowe) [inline test/helpers/deployZNS.ts]: let's keep "address" in the name, so we know this is not a contract object. we will call `accessController` the actual contract object. this way we know easily what is needed

File: test/helpers/deployZNS.ts [hunk 35]
```diff
@@ -134,20 +172,14 @@ export const deployZNS = async ({
...
-  zeroVaultAddress ?: string;
-}) : Promise<ZNSContracts> => {
+} : DeployZNSParams) : Promise<ZNSContracts> => {
   const accessController = await deployAccessController({
     deployer,
     governorAddresses: [deployer.address, ...governorAddresses],
     adminAddresses: [deployer.address, ...adminAddresses],
   });
 
+  // TODO AC: Make sure contracts are deployed as proxies and authorize the governor role
```
Comment (Whytecrowe) [inline test/helpers/deployZNS.ts]: do we still need this comment?

File: test/helpers/deployZNS.ts [hunk 36]
```diff
@@ -195,6 +227,11 @@ export const deployZNS = async ({
     registrar,
   };
 
+  // Final configuration steps
+  // TODO AC: remove all redundant calls here! and delete hashing of the root and the need
+  // for Registrar to be owner/operator of the root
+  await registry.connect(deployer).setOwnerOperator(registrar.address, true);
```
Comment (Whytecrowe) [inline test/helpers/deployZNS.ts]: why do we need this call here exactly? this seems to go against our AC structure.

File: test/helpers/types.ts [hunk 37]
```diff
@@ -6,11 +6,57 @@ import {
...
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
 
 export type Maybe<T> = T | undefined;
 
+export type GetterFunction =
+  (
+    Promise<string> |
+    Promise<boolean> |
+    Promise<BigNumber> |
+    Promise<Array<BigNumber>>
```
Comment (Whytecrowe) [inline test/helpers/types.ts]: would this be better if the type parameter is a union type rather than the parent type?

`Promise<string | boolean | ...>` ??

File: test/helpers/validateUpgrade.ts [hunk 38]
```diff
@@ -0,0 +1,57 @@
...
+  ] = await Promise.all(postUpgradeCalls);
+
+  // we're checking previous value after post upgrade calls
+  // to make sure that writing into new storage slots
+  // doesn't overwrite or corrupt the old ones
+  const postVals = await Promise.all(getters);
+
+  preVals.forEach((value, index) => {
+    expect(value).to.eq(postVals[index]);
+  });
```
Comment (Whytecrowe) [inline test/helpers/validateUpgrade.ts:51]: please, note what I changed here and why

**Review by Whytecrowe — CHANGES_REQUESTED** [review]
Please, address comments.
**Pull from origin before you do this, I've done a couple of commits here.**
Also, please, resolve a whole wagon of compiler warnings that appear when compiling on this branch. They all should be fixed.

File: contracts/access/AccessControlled.sol (hunk 10 above)
Comment (JamesEarle) [reply contracts/access/AccessControlled.sol]: maybe the change was lost somewhere in git when I merged everything, I didn't manually add this again. Will undo

File: contracts/distribution/ZNSEthRegistrar.sol (hunk 14 above)
Comment (JamesEarle) [reply contracts/distribution/ZNSEthRegistrar.sol]: unintentional

File: contracts/distribution/ZNSPriceOracle.sol (hunk 17 above)
Comment (JamesEarle) [reply contracts/distribution/ZNSPriceOracle.sol]: removed here

File: contracts/token/ZNSDomainToken.sol (hunk 22 above)
Comment (JamesEarle) [reply contracts/token/ZNSDomainToken.sol:12]: Where would you propose it be in this list? As long as they remain in the same order in future upgrades of this contract, it would be fine, no?

File: test/helpers/deployZNS.ts (hunk 36 above)
Comment (JamesEarle) [reply test/helpers/deployZNS.ts]: This was in place over a month ago before we had AC and when we still had subdomains. None of the contracts use this and the tests aren't reliant on it anymore so it was just redundant code, I removed it

File: test/helpers/deployZNS.ts (hunk 35 above)
Comment (JamesEarle) [reply test/helpers/deployZNS.ts]: nope, removed

File: test/ZNSDomainToken.test.ts (hunk 31 above)
Comment (JamesEarle) [reply test/ZNSDomainToken.test.ts]: fixed for this test
